File: front_end/app/utilities/board.ts
const colorOf = {
    Brick: 'rgb(170, 74, 68)',
    Desert: 'rgb(245, 213, 161)',
    Grain: 'rgb(250, 219, 94)',
    Ore: 'rgb(128, 128, 128)',
    Wood: 'rgb(34, 139, 34)',
    Wool: 'rgb(121, 208, 33)',
};

export const idToColor = {
    H01: colorOf.Ore,
    H02: colorOf.Wool,
    H03: colorOf.Wood,
    H04: colorOf.Grain,
    H05: colorOf.Brick,
    H06: colorOf.Wool,
    H07: colorOf.Brick,
    H08: colorOf.Grain,
    H09: colorOf.Wood,
    H10: colorOf.Desert,
    H11: colorOf.Wood,
    H12: colorOf.Ore,
    H13: colorOf.Wood,
    H14: colorOf.Ore,
    H15: colorOf.Grain,
    H16: colorOf.Wool,
    H17: colorOf.Brick,
    H18: colorOf.Grain,
    H19: colorOf.Wool,
} as const;

export type ID_Of_Hex = keyof typeof idToColor;

export const tokenMapping: { [key in ID_Of_Hex]: number | null } = {
    H01: 10,
    H02: 2,
    H03: 9,
    H04: 12,
    H05: 6,
    H06: 4,
    H07: 10,
    H08: 9,
    H09: 11,
    H10: null,
    H11: 3,
    H12: 8,
    H13: 8,
    H14: 3,
    H15: 4,
    H16: 5,
    H17: 5,
    H18: 6,
    H19: 11,
} as const;

const BOARD: ID_Of_Hex[][] = [
    ['H01', 'H02', 'H03'],
    ['H04', 'H05', 'H06', 'H07'],
    ['H08', 'H09', 'H10', 'H11', 'H12'],
    ['H13', 'H14', 'H15', 'H16'],
    ['H17', 'H18', 'H19']
];

const WIDTH_OF_BOARD_IN_VMIN = 100; // vmin
const NUMBER_OF_HEXES_THAT_SPAN_BOARD = 6;
const WIDTH_OF_HEX = WIDTH_OF_BOARD_IN_VMIN / NUMBER_OF_HEXES_THAT_SPAN_BOARD;
const RATIO_OF_LENGTH_OF_SIDE_OF_HEXAGON_AND_WIDTH_OF_HEXAGON = Math.tan(Math.PI / 6);
const RATIO_OF_HEIGHT_OF_HEX_AND_WIDTH_OF_HEX = 2 * RATIO_OF_LENGTH_OF_SIDE_OF_HEXAGON_AND_WIDTH_OF_HEXAGON;
const HEIGHT_OF_HEX = WIDTH_OF_HEX * RATIO_OF_HEIGHT_OF_HEX_AND_WIDTH_OF_HEX;
const DISTANCE_BETWEEN_BOTTOM_OF_HEX_IN_FIRST_ROW_AND_TOP_OF_HEX_IN_SECOND_ROW = HEIGHT_OF_HEX / 4;
const DISTANCE_BETWEEN_TOP_OF_HEX_IN_ONE_ROW_AND_TOP_OF_HEX_IN_NEXT_ROW = HEIGHT_OF_HEX - DISTANCE_BETWEEN_BOTTOM_OF_HEX_IN_FIRST_ROW_AND_TOP_OF_HEX_IN_SECOND_ROW;
const NUMBER_OF_ROWS_OF_HEXES_AFTER_FIRST = BOARD.length - 1;
const HEIGHT_OF_BOARD = HEIGHT_OF_HEX + DISTANCE_BETWEEN_TOP_OF_HEX_IN_ONE_ROW_AND_TOP_OF_HEX_IN_NEXT_ROW * NUMBER_OF_ROWS_OF_HEXES_AFTER_FIRST;
const VERTICAL_POSITION_OF_FIRST_ROW_OF_HEXES = (100 - HEIGHT_OF_BOARD) / 2; // vmin
const MARGIN_OF_ERROR = 0.01;

type Hex = { id: ID_Of_Hex; x: number; y: number };

export const hexes: Hex[] = BOARD.flatMap((row, index_of_row) => {
    const vertical_position_of_hex = VERTICAL_POSITION_OF_FIRST_ROW_OF_HEXES + index_of_row * DISTANCE_BETWEEN_TOP_OF_HEX_IN_ONE_ROW_AND_TOP_OF_HEX_IN_NEXT_ROW;
    const number_of_hexes_in_row = row.length;
    const horizontal_position_of_first_hex_in_row = (WIDTH_OF_BOARD_IN_VMIN - number_of_hexes_in_row * WIDTH_OF_HEX) / 2;
    return row.map((id_of_hex, index_of_hex) => ({
        id: id_of_hex,
        x: horizontal_position_of_first_hex_in_row + index_of_hex * WIDTH_OF_HEX,
        y: vertical_position_of_hex
    }));
});

type VertexCoord = { x: number; y: number };

function isClose(v1: VertexCoord, v2: VertexCoord, margin_of_error: number): boolean {
    return Math.abs(v1.x - v2.x) < margin_of_error && Math.abs(v1.y - v2.y) < margin_of_error;
}

export const vertices: VertexCoord[] = (() => {
    const uniqueVertices: VertexCoord[] = [];
    hexes.forEach(({ x, y }) => {
        const potentialVertices: VertexCoord[] = [
            { x: x + 0.5 * WIDTH_OF_HEX, y },
            { x: x + WIDTH_OF_HEX, y: y + 0.25 * HEIGHT_OF_HEX },
            { x: x + WIDTH_OF_HEX, y: y + 0.75 * HEIGHT_OF_HEX },
            { x: x + 0.5 * WIDTH_OF_HEX, y: y + HEIGHT_OF_HEX },
            { x, y: y + 0.75 * HEIGHT_OF_HEX },
            { x, y: y + 0.25 * HEIGHT_OF_HEX }
        ];
        potentialVertices.forEach(v => {
            if (!uniqueVertices.some(existing => isClose(existing, v, MARGIN_OF_ERROR))) {
                uniqueVertices.push(v);
            }
        })
    })
    return uniqueVertices;
})();

type Edge = { x1: number; y1: number; x2: number; y2: number };

function isEdgeClose(e1: Edge, e2: Edge, margin_of_error: number): boolean {
    const sameOrder =
        isClose({ x: e1.x1, y: e1.y1 }, { x: e2.x1, y: e2.y1 }, margin_of_error) &&
        isClose({ x: e1.x2, y: e1.y2 }, { x: e2.x2, y: e2.y2 }, margin_of_error);
    const swappedOrder =
        isClose({ x: e1.x1, y: e1.y1 }, { x: e2.x2, y: e2.y2 }, margin_of_error) &&
        isClose({ x: e1.x2, y: e1.y2 }, { x: e2.x1, y: e2.y1 }, margin_of_error);
    return sameOrder || swappedOrder;
}

export const edges: Edge[] = (() => {
    const edgesArray: Edge[] = [];
    hexes.forEach(({ x, y }) => {
        const vertices: VertexCoord[] = [
            { x: x + 0.5 * WIDTH_OF_HEX, y },
            { x: x + WIDTH_OF_HEX, y: y + 0.25 * HEIGHT_OF_HEX },
            { x: x + WIDTH_OF_HEX, y: y + 0.75 * HEIGHT_OF_HEX },
            { x: x + 0.5 * WIDTH_OF_HEX, y: y + HEIGHT_OF_HEX },
            { x, y: y + 0.75 * HEIGHT_OF_HEX },
            { x, y: y + 0.25 * HEIGHT_OF_HEX }
        ];
        for (let i = 0; i < vertices.length; i++) {
            const v1 = vertices[i];
            const v2 = vertices[(i + 1) % vertices.length];
            const newEdge: Edge = { x1: v1.x, y1: v1.y, x2: v2.x, y2: v2.y };
            if (!edgesArray.some(existing => isEdgeClose(existing, newEdge, MARGIN_OF_ERROR))) {
                edgesArray.push(newEdge);
            }
        }
    });
    return edgesArray;
})();

File: Python_back_end/board.py
from typing import List
from typing import Dict
from typing import Optional
from typing import Tuple
import json
import math
import numpy as np
from back_end.settings import settings


RATIO_OF_LENGTH_OF_SIDE_OF_HEX_AND_WIDTH_OF_HEX = math.tan(math.pi / 6)
TOKEN_DOT_MAPPING = {
    2: 1,
    3: 2,
    4: 3,
    5: 4,
    6: 5,
    8: 5,
    9: 4,
    10: 3,
    11: 2,
    12: 1
}
TOKEN_MAPPING = {
    "H01": 10,
    "H02": 2,
    "H03": 9,
    "H04": 12,
    "H05": 6,
    "H06": 4,
    "H07": 10,
    "H08": 9,
    "H09": 11,
    "H10": None,
    "H11": 3,
    "H12": 8,
    "H13": 8,
    "H14": 3,
    "H15": 4,
    "H16": 5,
    "H17": 5,
    "H18": 6,
    "H19": 11
}

RATIO_OF_HEIGHT_OF_HEX_AND_WIDTH_OF_HEX = 2 * RATIO_OF_LENGTH_OF_SIDE_OF_HEX_AND_WIDTH_OF_HEX
WIDTH_OF_HEX = settings.width_of_board_in_vmin / settings.number_of_hexes_that_span_board
HEIGHT_OF_HEX = WIDTH_OF_HEX * RATIO_OF_HEIGHT_OF_HEX_AND_WIDTH_OF_HEX
LENGTH_OF_SIDE_OF_HEX = WIDTH_OF_HEX * RATIO_OF_LENGTH_OF_SIDE_OF_HEX_AND_WIDTH_OF_HEX


def euclidean_distance_between_vertices_with_coordinates(x1: float, y1: float, x2: float, y2: float) -> float:
    return math.sqrt((x1 - x2)**2 + (y1 - y2)**2)


class Board:

    def __init__(self, geometry_file: Optional[str] = None):
        if geometry_file is None:
            geometry_file = settings.board_geometry_path
        with open(geometry_file, "r") as f:
            data = json.load(f)
        self.hexes: List[Dict] = data["hexes"]
        self.vertices: List[Dict] = data["vertices"]
        self.edges: List[Dict] = data["edges"]

        self._hex_vertex_cache: Dict[str, List[Tuple[float, float]]] = {}

        self._vertex_feature_map = {}
        for vertex in self.vertices:
            self._vertex_feature_map[vertex["label"]] = self.compute_vertex_feature(vertex)


    @staticmethod
    def get_edge_key(v1: Tuple[float, float], v2: Tuple[float, float]) -> str:
        (x1, y1) = v1
        (x2, y2) = v2
        # Sorts the two endpoints so that edge direction is irrelevant.
        if x1 < x2 or (x1 == x2 and y1 <= y2):
            return f"{x1:.2f}-{y1:.2f}_{x2:.2f}-{y2:.2f}"
        else:
            return f"{x2:.2f}-{y2:.2f}_{x1:.2f}-{y1:.2f}"


    def compute_vertex_feature(self, dictionary_of_vertex_information):
        x, y = dictionary_of_vertex_information["x"], dictionary_of_vertex_information["y"]
        total_number_of_pips = 0
        number_of_hexes = 0
        array_of_coordinates = np.array([x, y])
        for dictionary_of_hex_information in self.hexes:
            array_of_coordinates_of_vertices_of_hex = np.array(self.get_hex_vertices(dictionary_of_hex_information))
            if np.any(
                np.all(
                    np.abs(array_of_coordinates_of_vertices_of_hex - array_of_coordinates) < settings.margin_of_error,
                    axis = 1
                )
            ):
                id_of_hex = dictionary_of_hex_information["id"]
                number_of_token = TOKEN_MAPPING.get(id_of_hex)
                if number_of_token is not None:
                    number_of_pips = TOKEN_DOT_MAPPING.get(number_of_token, 0)
                    total_number_of_pips += number_of_pips
                    number_of_hexes += 1
        normalized_total_number_of_pips = total_number_of_pips / (number_of_hexes * 5) if number_of_hexes > 0 else 0.0
        normalized_x_coordinate = x / settings.width_of_board_in_vmin
        normalized_y_coordinate = y / 100.0
        normalized_number_of_hexes = number_of_hexes / 3.0
        return [normalized_total_number_of_pips, normalized_x_coordinate, normalized_y_coordinate, normalized_number_of_hexes, 1.0]


    def get_hex_vertices(self, hex_tile: Dict) -> List[Tuple[float, float]]:
        '''
        Given a dictionary of hex information, return a list of pairs of coordinates of the vertices of the hex.
        This function caches the result so that repeated calls with the same hex (determined by its ID)
        do not recompute the vertex coordinates.
        '''
        hex_id = hex_tile.get("id")
        if hex_id in self._hex_vertex_cache:
            return self._hex_vertex_cache[hex_id]

        x = hex_tile["x"]
        y = hex_tile["y"]
        vertices = [
            (x + 0.5 * WIDTH_OF_HEX, y),
            (x + WIDTH_OF_HEX, y + 0.25 * HEIGHT_OF_HEX),
            (x + WIDTH_OF_HEX, y + 0.75 * HEIGHT_OF_HEX),
            (x + 0.5 * WIDTH_OF_HEX, y + HEIGHT_OF_HEX),
            (x, y + 0.75 * HEIGHT_OF_HEX),
            (x, y + 0.25 * HEIGHT_OF_HEX)
        ]
        self._hex_vertex_cache[hex_id] = vertices
        return vertices


    def get_vertex_by_label(self, label: str) -> Optional[Dict]:
        for vertex in self.vertices:
            if vertex["label"] == label:
                return vertex
        return None


    def get_vertex_features(self, vertex_label: str) -> Optional[List[float]]:
        '''
        Return the precomputed feature vector for the given vertex label.
        '''
        return self._vertex_feature_map.get(vertex_label)


    def get_available_building_moves(self, list_of_labels_of_occupied_vertices: List[str]) -> List[str]:
        '''
        Return the list of labels of vertices that are not too close to any vertex in the given list of labels of occupied vertices.
        Two vertices are considered too close if the Euclidean distance between them is less than the length of the side of a hex plus a small margin.
        '''
        list_of_tuples_of_coordinates_of_occupied_vertices = []
        for label in list_of_labels_of_occupied_vertices:
            dictionary_of_vertex_information = self.get_vertex_by_label(label)
            if dictionary_of_vertex_information:
                x = dictionary_of_vertex_information["x"]
                y = dictionary_of_vertex_information["y"]
                tuple_of_coordinates_of_occupied_vertex = (x, y)
                list_of_tuples_of_coordinates_of_occupied_vertices.append(tuple_of_coordinates_of_occupied_vertex)
        list_of_labels_of_available_vertices = []
        for dictionary_of_vertex_information in self.vertices:
            label = dictionary_of_vertex_information["label"]
            if label in list_of_labels_of_occupied_vertices:
                continue
            x1 = dictionary_of_vertex_information["x"]
            y1 = dictionary_of_vertex_information["y"]
            if any(
                euclidean_distance_between_vertices_with_coordinates(x1, y1, x2, y2) < (LENGTH_OF_SIDE_OF_HEX + settings.margin_of_error)
                for x2, y2 in list_of_tuples_of_coordinates_of_occupied_vertices
            ):
                continue
            list_of_labels_of_available_vertices.append(label)
        return list_of_labels_of_available_vertices


    def get_available_road_moves(self, last_settlement_or_city: str, used_edge_keys: List[str]) -> List[Tuple[Dict, str]]:
        '''
        Return a list of (edge, edge_key) tuples adjacent to the given settlement or city.
        '''
        vertex = self.get_vertex_by_label(last_settlement_or_city)
        if vertex is None:
            return []
        settlement_or_city_coord = (vertex["x"], vertex["y"])
        adjacent_edges = []
        for edge in self.edges:
            v1 = (edge["x1"], edge["y1"])
            v2 = (edge["x2"], edge["y2"])
            if (
                (math.isclose(v1[0], settlement_or_city_coord[0], abs_tol = settings.margin_of_error) and math.isclose(v1[1], settlement_or_city_coord[1], abs_tol = settings.margin_of_error)) or
                (math.isclose(v2[0], settlement_or_city_coord[0], abs_tol = settings.margin_of_error) and math.isclose(v2[1], settlement_or_city_coord[1], abs_tol = settings.margin_of_error))
            ):
                edge_key = self.get_edge_key(v1, v2)
                if edge_key not in used_edge_keys:
                    adjacent_edges.append((edge, edge_key))
        return adjacent_edges